src/test/parallel.py

```diff
@@ -0,0 +1,864 @@
+              if test_group in self.running_groups:
+                # Try to find other non-running test group.
+                continue
+              else:
+                self.running_groups.add(test_group)
+
+            del self.tasks[task_id]
+            break
+          else:
+            # Either there is no tasks left or number or remaining test
+            # cases (groups) is less than number or running threads.
+            return
+
+        task_manager.run_task(task)
+
+        if self.running_groups is not None:
+          with self.task_lock:
+            self.running_groups.remove(test_group)
+
+  def start_daemon(func):
+    t = threading.Thread(target=func)
+    t.daemon = True
+    t.start()
+    return t
+
+  try:
+    if timeout:
+      timeout.start()
+    running_groups = set() if serialize_test_cases else None
+    worker_fn = WorkerFn(tasks, running_groups)
+    workers = [start_daemon(worker_fn) for _ in range(pool_size)]
+    for worker in workers:
+      worker.join()
+  finally:
+    if timeout:
+      timeout.cancel()
+
+
+def default_options_parser(default_binary, file_test_list):
+  parser = optparse.OptionParser(
+      usage = 'usage: %prog [options] binary [binary ...] -- [additional args]')
+
+  parser.add_option('-d', '--output_dir', type='string', default=None,
+                    help='Output directory for test logs. Logs will be '
+                         'available under gtest-parallel-logs/, so '
+                         '--output_dir=/tmp will results in all logs being '
+                         'available under /tmp/gtest-parallel-logs/.')
+  parser.add_option('-r', '--repeat', type='int', default=1,
+                    help='Number of times to execute all the tests.')
+  parser.add_option('--retry_failed', type='int', default=0,
+                    help='Number of times to repeat failed tests.')
+  parser.add_option('--failed', action='store_true', default=False,
+                    help='run only failed and new tests')
+  parser.add_option('-w', '--workers', type='int',
+                    default=multiprocessing.cpu_count(),
+                    help='number of workers to spawn')
+  parser.add_option('--gtest_color', type='string', default='yes',
+                    help='color output')
+  parser.add_option('--gtest_filter', type='string', default='',
+                    help='test filter')
+  parser.add_option('--gtest_also_run_disabled_tests', action='store_true',
+                    default=False, help='run disabled tests too')
+  parser.add_option('--print_test_times', action='store_true', default=False,
+                    help='list the run time of each test at the end of execution')
+  parser.add_option('--shard_count', type='int', default=1,
+                    help='total number of shards (for sharding test execution '
+                         'between multiple machines)')
+  parser.add_option('--shard_index', type='int', default=0,
```

Comment:
Not sure it makes sense to give `shard_index` a default. I think you want to ensure that `shard_count` and `shard_index` are used in combination, so if `shard_count` is used and `options.shard_index` is `None`,  you print proper usage. Right now `options.shard_index` will just default to 0, so you can't tell if it's being used properly with `shard_count`.